SAK-32296 Add NYU's "auto favorites" feature

marktriggs opened the pull request:
There are a few components to this:

  * The "Organize Favorites" tab of the "Sites" drawer now shows an
    "Automatically add new sites to your favorites bar" toggle.  If
    this is enabled (the default), user favorites are automatically
    added to include sites as they appear.

  * To avoid the user being overwhelmed by sites, we introduce a limit
    of 15 sites (configurable via sakai.properties) to the site bar at
    the top of the portal.  The "Organize Favorites" tab shows a warning
    if the user has more than this many sites, and provides a visual
    indication of which sites are above the fold.

  * Since users will want to occasionally prune their list of favorite
    sites, there's a "select all/select none" star at the top of each
    term shown in the Sites drawer.  This allows all sites within a term
    to be favorited or unfavorited at once.

  * The "Reload to see your updated favorites" indicator will now only
    show if the user has made a change.  I.e. favoriting and then
    unfavoriting a site won't show the indicator.

master-bob commented on portal/portal-impl/impl/pom.xml:
```diff
@@ -114,6 +114,11 @@
         <dependency>
+            <groupId>com.googlecode.json-simple</groupId>
+            <artifactId>json-simple</artifactId>
+            <version>1.1.1</version>
```
Use ${json.simple.version}

marktriggs commented:
Oops, quite right.  Will fix this soon.

marktriggs commented:
Hi there,

I've i18n'd the places that needed it, squashed the commits, rebased off master and linked it to the JIRA (that I only just discovered ;).  Ready for review, thanks!

ottenhoff commented on portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/handlers/FavoritesHandler.java:
```diff
@@ -105,10 +188,44 @@ private void setFavoriteSiteIds(String userId, String[] siteIds) throws Permissi
+		public static UserFavorites fromJSON(String json) throws ParseException {
+			JSONParser parser = new JSONParser();
+
+			JSONObject obj = (JSONObject)parser.parse(json);
```
I'm getting an NPE right here in my testing

marktriggs replied:
Thanks Sam.  @payten just saw this too, but it was a JS caching issue (which I guess shouldn't happen outside of development if people are using `portal.cdn.version`).  Does a shift-refresh fix yours?

ottenhoff replied:
Yes, this works well after a shift-refresh.

master-bob replied:
How will some of the other exceptions that may occur be handled? That is what if a ClassCastException occurs (say if a JSONArray is returned instead of a JSONObject.

marktriggs replied:
For this to happen, the JSON coming from the user's browser (via an AJAX POST request) would have to be invalid.  If that's the case (i.e. the JSON got edited manually, truncated or was missing entirely), we wrap the exception in a ServletException and return a 500.  The calling JavaScript will abort the update and refresh the browser's state from the server.

So basically the policy is "if anything goes wrong with parsing user input, we throw an exception and bail out".
